module/state_synchronization/storage.go

```diff
@@ -0,0 +1,957 @@
+		return err
+	}
+
+	return nil
+}
+
+func (s *Storage) bootstrap(startHeight uint64) error {
+	latestIncorporatedHeightKey := makeGlobalStateKey(globalStateLatestIncorporatedHeight)
+	latestIncorporatedHeightValue := make([]byte, 8)
+	binary.LittleEndian.PutUint64(latestIncorporatedHeightValue, startHeight)
+
+	storedDataLowerBoundKey := makeGlobalStateKey(globalStateStoredDataLowerBound)
+	storedDataLowerBoundValue := make([]byte, 8)
+	binary.LittleEndian.PutUint64(storedDataLowerBoundValue, startHeight)
+
+	return s.db.Update(func(txn *badger.Txn) error {
+		if err := txn.Set(latestIncorporatedHeightKey, latestIncorporatedHeightValue); err != nil {
+			return fmt.Errorf("failed to set latest incorporated height: %w", err)
+		}
+
+		if err := txn.Set(storedDataLowerBoundKey, storedDataLowerBoundValue); err != nil {
+			return fmt.Errorf("failed to set stored data lower bound: %w", err)
+		}
+
+		return nil
+	})
+}
+
+func (s *Storage) GetLatestIncorporatedHeight() (uint64, error) {
+	latestIncorporatedHeightKey := makeGlobalStateKey(globalStateLatestIncorporatedHeight)
+	var latestIncorporatedHeight uint64
+
+	if err := s.db.View(func(txn *badger.Txn) error {
+		item, err := txn.Get(latestIncorporatedHeightKey)
+		if err != nil {
+			return fmt.Errorf("failed to find latest incorporated height entry: %w", err)
+		}
+
+		latestIncorporatedHeightValue, err := item.ValueCopy(nil)
+		if err != nil {
+			return fmt.Errorf("failed to retrieve latest incorporated height value: %w", err)
+		}
+
+		latestIncorporatedHeight = binary.LittleEndian.Uint64(latestIncorporatedHeightValue)
+
+		return nil
+	}); err != nil {
+		return 0, err
+	}
+
+	return latestIncorporatedHeight, nil
+}
+
+// GetStoredDataLowerBound returns the maximum h such that the storage contains no data for any block heights <= h
+// TODO: please help me think of a better name for this function
+func (s *Storage) GetStoredDataLowerBound() (uint64, error) {
+	storedDataLowerBoundKey := makeGlobalStateKey(globalStateStoredDataLowerBound)
+	var storedDataLowerBound uint64
+
+	if err := s.db.View(func(txn *badger.Txn) error {
+		item, err := txn.Get(storedDataLowerBoundKey)
+		if err != nil {
+			return fmt.Errorf("failed to find stored data lower bound entry: %w", err)
+		}
+
+		storedDataLowerBoundValue, err := item.ValueCopy(nil)
+		if err != nil {
+			return fmt.Errorf("failed to retrieve stored data lower bound value: %w", err)
+		}
+
+		storedDataLowerBound = binary.LittleEndian.Uint64(storedDataLowerBoundValue)
+
+		return nil
+	}); err != nil {
+		return 0, err
+	}
+
+	return storedDataLowerBound, nil
+}
+
+// Prune removes all data from storage corresponding to block heights lower than or equal to the given height.
+// This includes all blob tree records, blob records, and blob data.
+// This should only be called with height <= latest incorporated height.
+func (s *Storage) Prune(height uint64) ([]cid.Cid, error) {
```

Comment:
### Abstraction
I think this storage module lacks of levels of abstraction. The exposed implementation details makes the main logic not that easy to understand.

We need provide abstraction for database operation, so that the logic here could work with higher level abstraction. 

As an example. 

I think the logic could be more clear with higher level abstractions: (pseudo code)
```
func (s) PruneUpToHeight(height) ([]cid, error) {
  pruned := s.executionDataSyncTracker.GetPrunedHeight(s.db)
  
  if height <= pruned {
    // preventing the pruned height go backwards
    return nil
  }
  
  for h := pruned + 1; h <= height; h++ {
    err := s.pruneByHeight(height)
  }

  // only update after the operation is done, in case it crashed before updating db,
  // we will prune again after restart, and the pruning operation is a no-op if has done before.
  s.executionDataSyncTracker.SetPrunedHeight(s.db, height)
  return nil
}
```

The encoding from uint64 to bytes should only stay in the low level DB operation implementation.

The low level database operation should provide abstraction for getting and setting key value pairs with the most convenient data type for the upper level application, where as the low level deals with encoding.

### Data Type
We are storing multiple simple data type under different keys with one transaction. I wonder if makes sense to construct a structured object and [encoding it with msgpack](https://github.com/onflow/flow-go/blob/master/storage/badger/operation/common.go#L73) and store it under a single key.

Reply:
Thanks, I agree with this, but the storage module is actually intended to be a low level API. It is only used by the Adder / Getter, and all other code should only use the Adder and Getter. Does this make sense?

Reply:
> We are storing multiple simple data type under different keys with one transaction. I wonder if makes sense to construct a structured object and [encoding it with msgpack](https://github.com/onflow/flow-go/blob/master/storage/badger/operation/common.go#L73) and store it under a single key.

I would have liked to do this for blob records and blob data, but the blob data is currently being populated by the blockservice third party library, and it would be too much work to mess with the implementation of that.